subiquity/models/ad.py

```diff
@@ -0,0 +1,33 @@
+# Copyright 2023 Canonical, Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+from typing import Optional
+
+from subiquity.common.types import ADConnectionInfo
+
+log = logging.getLogger('subiquity.models.ad')
+
+
+class ADModel:
+    """ Models the Active Directory feature """
+    def __init__(self) -> None:
+        self.do_join = False
+        self.conn_info: Optional[ADConnectionInfo] = None
+
+    async def target_packages(self):
```

Comment:
What's the plan for this?  Curtin does have some logic here for carrying a list of needed packages and it dynamically installs them if needed.  Would you expect them to be on the system already, for both the minimal and non source cases?

Reply:
Those packages must be on the live CD pool. Lots of users won't join AD, thus it makes sense for not having them inthe minimal layer. Having them on the pool enables curtin to install them in the target system for those that opt in for AD.

Reply:
Having the packages in the ship-live pool is just fine and matches other similar things.  We just might want to move this list to curtin.

Reply:
Is there any other endpoint that moved a similar list to curtin we could follow the trend on? I've seen the pattern herein applied in a few places in subiquity's source. It might be worth doing them altoghether if the procedure is straightforward.

Reply:
I think if the logic for joining the domain etc is going into curtin, the list of package required should too, probably.

Reply:
curtin has a `detect_required_packages` concept, and I look at this as very similar.  But maybe that particular detection is too late.

Reply:
My plan A is to follow the example of codecs, ssh and network, where we keep a list of optional packages in the model to get them installed in the target system by the install controller's postinstall method (if necessary):

https://github.com/canonical/subiquity/blob/f77971238f179cbcb0db8b16279157ef6801db28/subiquity/server/controllers/install.py#L382-L394

I don't think that fits in curtin (as Dan suggested) because we consider those packages optional and can only now they are required when the user interacts with the client UI.

Reply:
I might be wrong about curtin though. My understanding is that `detect_required_packages` output depends on the configuration data, not matching the interactive use case.

Reply:
Unless we install the packages as part of the same action (command) used to join the domain. Does that sound as awkward to you as to me? I mean, the command would have the responsibility of installing packages and performing the join. Besides SRP, it feels strange to have distinct moments during the installation when we install packages. It seems more natural do have one single moment of `Installing additional packages...`.

I'm not sure. Let me know your thoughts.